rpc: Replace median fee rate with feerate percentiles in getblockstats

marcinja opened the pull request:
Currently,  the `medianfeerate` statistic is calculated from the feerate of the middle transaction of a list of transactions sorted by feerate. 

This PR instead uses the value of the 50th percentile weight unit in the block, and also calculates the feerate at the 10th, 25th, 75th, and 90th percentiles.  This more accurately corresponds with what is generally meant by median feerate.

Commit:
Replace median fee rate with feerate percentiles

Removes medianfeerate result from getblockstats.
Adds feerate_percentiles which give the feerate of the 10th, 25th, 50th,
75th, and 90th percentile weight unit in the block.

jnewbery commented:
Concept ACK. The feerate is a measure on weight unit, not discrete transaction, so the median feerate should me the middle weight unit of the block.

Conceptually, this gives a better representation of the fees paid in the block, since the measure is not skewed by small or large txs.

Will review tomorrow.

MarcoFalke commented:
utACK 2d4bf95973bb5224238762ce94e3ecb835b77280

MarcoFalke commented:
Imo this is a bugfix for 0.17, but not a blocker for rc1.

ajtowns commented on src/rpc/blockchain.cpp:
```diff
@@ -1640,6 +1640,37 @@ static T CalculateTruncatedMedian(std::vector<T>& scores)
+    int64_t cumulative_weight = 0;
+    for (const auto& element: scores) {
+        cumulative_weight += element.second;
+        while (cumulative_weight >= weights[next_percentile_index]) {
```
This should be `while (next_percentile_index < NUM_GETBLOCKSTATS_PERCENTILS && cumulative_weight >= weights[next_percentile_index]) {`, no?

marcinja replied:
Yes, thanks for catching that.

w7788129 commented on src/rpc/blockchain.cpp:
```diff
@@ -1640,6 +1640,37 @@ static T CalculateTruncatedMedian(std::vector<T>& scores)
+{
+    const size_t size = scores.size();
+    if (size == 0) {
+        return;
```
'1HD2DVx2oSMiXPHX5XWXXBnwQnKSEZdEuY'

kallewoof commented on src/rpc/blockchain.cpp:
```diff
@@ -1679,7 +1710,13 @@ static UniValue getblockstats(const JSONRPCRequest& request)
             "  \"maxtxsize\": xxxxx,       (numeric) Maximum transaction size\n"
             "  \"medianfee\": xxxxx,       (numeric) Truncated median fee in the block\n"
-            "  \"medianfeerate\": xxxxx,   (numeric) Truncated median feerate (in satoshis per virtual byte)\n"
+            "  \"feerate_percentiles\": [   (array of numeric) Feerates at the 10th, 25th, 50th, 75th, and 90th percentile weight unit (in satoshis per virtual byte)\n"
```
Nit: indentation is off for `(array of numeric) [...]` -- see above line.

jnewbery replied:
nit: the other result keys are dictionary sorted in the help text. I suggest you follow the same convention here.

kallewoof reviewed (commented):
utACK 1fd565be86e3727e8dabcb961f9a24c18bddd2a8

domob1812 commented on src/rpc/blockchain.cpp:
```diff
@@ -1640,6 +1640,37 @@ static T CalculateTruncatedMedian(std::vector<T>& scores)
+static void CalculatePercentilesByWeight(T result[NUM_GETBLOCKSTATS_PERCENTILES], std::vector<std::pair<T, int64_t>>& scores, int64_t total_weight)
+{
+    const size_t size = scores.size();
+    if (size == 0) {
```
You don't use `size` anywhere else below - so you could just do:

    if (scores.empty()) {
        return;
    }

jnewbery replied:
:+1:

domob1812 commented on src/rpc/blockchain.cpp:
```diff
@@ -1640,6 +1640,37 @@ static T CalculateTruncatedMedian(std::vector<T>& scores)
+    };
+    int64_t next_percentile_index = 0;
+    int64_t cumulative_weight = 0;
+    for (const auto& element: scores) {
```
Nit: I believe there should be spaces on both sides of `:`.

conscott commented on src/rpc/blockchain.cpp:
```diff
@@ -1640,6 +1640,37 @@ static T CalculateTruncatedMedian(std::vector<T>& scores)
+    int64_t cumulative_weight = 0;
+    for (const auto& element: scores) {
+        cumulative_weight += element.second;
+        while (next_percentile_index < NUM_GETBLOCKSTATS_PERCENTILES && cumulative_weight >= weights[next_percentile_index]) {
```
This logic seems correct, but would be nice to have some test cases to ensure expected behavior for cases like:
1. Few number of txs (expect same feerate for multiple percentiles)
2. Tx whose weight spans more than one weight percentile group

jnewbery replied:
Agree that `getblockstats` could do with some more thorough testing, but this PR doesn't make things worse, so I'd be inclined to say leave it for a future PR.

conscott reviewed (commented):
utACK 1fd565be86e3727e8dabcb961f9a24c18bddd2a8

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1640,6 +1640,37 @@ static T CalculateTruncatedMedian(std::vector<T>& scores)
     }
 }
 
+static constexpr int NUM_GETBLOCKSTATS_PERCENTILES = 5;
```
nit: I think it's better to place `constexpr`s at the top of the file (especially since this is used in multiple functions in this file)

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1640,6 +1640,37 @@ static T CalculateTruncatedMedian(std::vector<T>& scores)
 
+static constexpr int NUM_GETBLOCKSTATS_PERCENTILES = 5;
+
+template<typename T>
```
nit: using a template here is over-engineering this. Even if you plan to re-use this a measure other than fee-rate, it would make more sense to define this just for fee-rate now and then change to a template in the next PR.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1640,6 +1640,37 @@ static T CalculateTruncatedMedian(std::vector<T>& scores)
+        }
+    }
+
+    for (int64_t i = next_percentile_index; i < NUM_GETBLOCKSTATS_PERCENTILES; i++) {
```
nit: perhaps a comment here to say that you're filling in the remaining percentile results to the same as the last filled percentile result.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1867,7 +1907,13 @@ static UniValue getblockstats(const JSONRPCRequest& request)
     ret_all.pushKV("medianfee", CalculateTruncatedMedian(fee_array));
-    ret_all.pushKV("medianfeerate", CalculateTruncatedMedian(feerate_array));
+
+    UniValue feerates_res(UniValue::VARR);
```
nit: I'd push this code up to immediately below the `CalculatePercentilesByWeight()` call so all the `ret_all.pushKV()` calls are on contiguous (alphabetized) lines.

jnewbery commented on test/functional/data/rpc_getblockstats.json:
```diff
@@ -174,13 +186,19 @@
       "avgfeerate": 109,
       "avgtxsize": 228,
       "blockhash": "22d9b8b9c2a37c81515f3fc84f7241f6c07dbcea85ef16b00bcc33ae400a030f",
+	"feerate_percentiles": [
```
nit: replace tab with spaces please :)

jnewbery reviewed (commented):
Code looks correct so utACK 1fd565be86e3727e8dabcb961f9a24c18bddd2a8

I've also got a bunch of nits for you :)

marcinja commented:
Thanks for the reviews. I fixed all the nits and added some unit tests for `CalculatePercentilesByWeight`

I'm not sure that I put the tests in the best location, and also this now fails on Travis because of linting

```
✗ bitcoin (change-feerates-rpc) ./test/lint/lint-includes.sh 
The following files #include .cpp files:
src/test/rpc_tests.cpp:#include <rpc/blockchain.cpp>
```

MarcoFalke commented on src/test/rpc_tests.cpp:
```diff
@@ -16,6 +16,8 @@
 
 #include <univalue.h>
 
+#include <rpc/blockchain.cpp>
```
Can't you just include the .h file?

MarcoFalke replied:
If there are static methods that you need, remove the static in the cpp file and add the declaration here

marcinja replied:
Yup, that works thanks. I had to move some of the other includes but I think that should be okay.

domob1812 commented on src/rpc/blockchain.cpp:
```diff
@@ -1848,26 +1885,34 @@ static UniValue getblockstats(const JSONRPCRequest& request)
-            if (do_medianfeerate) {
-                feerate_array.push_back(feerate);
+            if (do_feerate_percentiles) {
+                feerate_array.emplace_back(std::make_pair(feerate, weight));
```
With `emplace_back`, I think you should be able to get rid of `make_pair` - `std::pair` has a constructor with `first` and `second`, so that should be called then.

marcinja replied:
So use `std::pair<CAmount, int64_t>(feerate, weight)` instead? Is there a reason the constructor is preferred?

MarcoFalke replied:
Unless there is compelling reason to change the code, you should keep the it as is for now.

domob1812 replied:
No, I meant to just use `freerate_array.emplace_back(feerate, weight)`.  I haven't tested it, but according to my understanding it should work - and just be simpler in the code.  Whether that is a compelling reason or not, I don't know.

kallewoof replied:
Agree with @domob1812 that
```C++
feerate_array.emplace_back(feerate, weight);
```
looks cleaner (and yes, it works).

domob1812 reviewed (commented):
utACK b4718d26ab785229ecfb7d89b593cb050d883fc3

achow101 commented:
utACK 64101856b4e9a110770e389d2a1cad6a5802321c

sipa commented:
This breaks the RPC interface. Do we need to go through the rpc deprecation process, or should we should maintain a legacy `median_feerate` field?

achow101 commented:
@sipa getblockstats is not in any release, so I think this is actually fine to have a breaking API change.

sipa commented:
Ah, I didn't notice this was considered to be a bugfix.

Empact commented on src/rpc/blockchain.cpp:
```diff
@@ -32,7 +31,6 @@
 #include <warnings.h>
 
 #include <assert.h>
-#include <stdint.h>
```
nit: you should leave these includes
> Every .cpp and .h file should #include every header file it directly uses classes, functions or other definitions from, even if those headers are already included indirectly through other headers.
https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#source-code-organization

marcinja replied:
Fixed.

MarcoFalke commented:
re-utACK 4b7091a842c2c3a76f4136cb0fdcf1c5904fd237

kallewoof reviewed (commented):
utACK 4b7091a842c2c3a76f4136cb0fdcf1c5904fd237

laanwj commented:
Cannot really test this without txindex enabled,
utACK 4b7091a842c2c3a76f4136cb0fdcf1c5904fd237
